orttraining/orttraining/training_ops/cuda/nn/batch_norm_grad.cc

```diff
@@ -5,86 +5,131 @@
   const TensorShape input_shape = X->Shape();
   const TensorShape channel_shape = saved_mean->Shape();
 
   // no B here, but B has same size as Scale, so can validate inputs for gradient with this substitute
-  ORT_RETURN_IF_ERROR(BatchNormHelper::ValidateInputs(X, Scale, Scale, saved_mean, saved_variance));
+  ORT_RETURN_IF_ERROR(BatchNormHelper::ValidateInputs(X, Scale, Scale, saved_mean, saved_inv_std));
 
   auto dY_data = reinterpret_cast<const CudaT*>(dY->template Data<T>());
   auto X_data = reinterpret_cast<const CudaT*>(X->template Data<T>());
-  auto Scale_data = reinterpret_cast<const CudaT*>(Scale->template Data<T>());
-  auto saved_mean_data = reinterpret_cast<const CudaT*>(saved_mean->template Data<T>());
-  auto saved_variance_data = reinterpret_cast<const CudaT*>(saved_variance->template Data<T>());
+  auto Scale_data = reinterpret_cast<const CudaT1*>(Scale->template Data<T1>());
+  auto saved_mean_data = reinterpret_cast<const CudaT2*>(saved_mean->template Data<T2>());
+  auto saved_inv_std_data = reinterpret_cast<const CudaT2*>(saved_inv_std->template Data<T2>());
 
   auto dX_data = reinterpret_cast<CudaT*>(ctx->Output(0, input_shape)->template MutableData<T>());
-  auto dScale_data = reinterpret_cast<CudaT*>(ctx->Output(1, channel_shape)->template MutableData<T>());
-  auto dBias_data = reinterpret_cast<CudaT*>(ctx->Output(2, channel_shape)->template MutableData<T>());
+  auto dScale_data = reinterpret_cast<CudaT1*>(ctx->Output(1, channel_shape)->template MutableData<T1>());
+  auto dBias_data = reinterpret_cast<CudaT1*>(ctx->Output(2, channel_shape)->template MutableData<T1>());
 
-  const auto alpha = Consts<CudaT>::One;
-  const auto beta = Consts<CudaT>::Zero;
+  const auto alpha = std::is_same<T, MLFloat16>::value ? Consts<float>::One : Consts<CudaT>::One;
+  const auto beta = std::is_same<T, MLFloat16>::value ? Consts<float>::Zero : Consts<CudaT>::Zero;
 
   CudnnTensor input_tensor, scale_bias_tensor;
   vector<int64_t> new_dims;
   BatchNormHelper::NormalizeDims(input_shape, new_dims);
   ORT_RETURN_IF_ERROR(input_tensor.Set(new_dims, CudnnTensor::GetDataType<CudaT>()));
+  // for fp16 input, `scale_bias_tensor` will have a float type; otherwise it will be the same as input type.
   ORT_RETURN_IF_ERROR(scale_bias_tensor.Set(input_tensor, cudnn_batch_norm_mode_));
 
-  // note this is only valid for cudnnBatchNormalizationForwardTraining, not ForwardInference
-  CUDNN_RETURN_IF_ERROR(
-      cudnnBatchNormalizationBackward(
-          CudnnHandle(),
-          cudnn_batch_norm_mode_,
-          &alpha,
-          &beta,
-          &alpha,
-          &beta,
-          input_tensor,
-          X_data,
-          input_tensor,
-          dY_data,
-          input_tensor,
-          dX_data,
-          scale_bias_tensor,
-          Scale_data,
-          dScale_data,
-          dBias_data,
-          epsilon_,
-          saved_mean_data,
-          saved_variance_data));
+  const int64_t C = input_shape.GetDims()[1];
```

Comment:
According to spec: "The op also accepts single dimension input of size N in which case C is assumed to be 1", so using this way to get the C value seems not always right. Maybe also need more UT to cover such case?

Actually in BatchNormHelper::ValidateInputs we require dim of X > 2, seems this is not following the spec.

Reply:
Corrected this. Also noted that the Onnx spec contradicts itself too: [doc](https://github.com/onnx/onnx/blob/31f0fd6103e231664d1cfd3b25264f672f5de960/onnx/defs/nn/defs.cc#L1663) and [def](https://github.com/onnx/onnx/blob/31f0fd6103e231664d1cfd3b25264f672f5de960/onnx/defs/nn/defs.cc#L1747). Its def requires the dim size of input X greater than 1. Maybe the spec should also be changed to keep consistent.

Reply:
Raised [an issue](https://github.com/onnx/onnx/issues/3555) related to Onnx spec contradiction.